onnx/defs/shape_inference.h

```diff
@@ -74,11 +74,28 @@ struct InferenceContext {
   virtual const SparseTensorProto* getInputSparseData(size_t index) const = 0;
 };
 
+struct DataPropagationContext {
+  virtual const AttributeProto* getAttribute(const std::string& name) const = 0;
+  virtual size_t getNumInputs() const = 0;
+  virtual const TypeProto* getInputType(size_t index) const = 0;
+  virtual size_t getNumOutputs() const = 0;
+  virtual TypeProto* getOutputType(size_t index) = 0;
+  virtual std::string getOutputName(size_t index) const = 0;
```

Comment:
when do we need output name? and what about input name? do we need that too?

Reply:
Agree. I don't think it needs to be in this interface. (Not needed for the node-level propagation function, even though the *implementation* may use the names.)